pkg/frontend/querymiddleware/error_caching.go

```diff
@@ -0,0 +1,182 @@
+
+	cacheAttempted    prometheus.Counter
+	cacheHits         prometheus.Counter
+	cacheStoreSkipped *prometheus.CounterVec
+
+	// TODO: Cache store attempted metric? skipped vs stored?
+}
+
+func (e *errorCachingHandler) Do(ctx context.Context, request MetricsQueryRequest) (Response, error) {
+	spanLog := spanlogger.FromContext(ctx, e.logger)
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return e.next.Do(ctx, request)
+	}
+
+	// Check if caching has disabled via an option on the request
+	if !e.shouldCacheReq(request) {
+		e.cacheStoreSkipped.WithLabelValues(reasonDisabledByOption).Inc()
+		return e.next.Do(ctx, request)
+	}
+
+	e.cacheAttempted.Inc()
+	key := e.cacheKey(tenant.JoinTenantIDs(tenantIDs), request)
+	hashedKey := cacheHashKey(key)
+
+	if cachedErr := e.loadErrorFromCache(ctx, key, hashedKey); cachedErr != nil {
+		e.cacheHits.Inc()
+		spanLog.DebugLog(
+			"msg", "returned cached API error",
+			"error_type", cachedErr.Type,
+			"key", key,
+			"hashed_key", hashedKey,
+		)
+
+		return nil, cachedErr
+	}
+
+	res, err := e.next.Do(ctx, request)
+	if err != nil {
+		var apiErr *apierror.APIError
+		if !errors.As(err, &apiErr) {
+			e.cacheStoreSkipped.WithLabelValues(reasonNotAPIError).Inc()
+			return res, err
+		}
+
+		if cacheable, reason := e.isCacheable(apiErr, request, tenantIDs); !cacheable {
+			e.cacheStoreSkipped.WithLabelValues(reason).Inc()
+			spanLog.DebugLog(
+				"msg", "error result from request is not cacheable",
+				"error_type", apiErr.Type,
+				"reason", reason,
+			)
+			return res, err
+		}
+
+		e.storeErrorToCache(key, hashedKey, apiErr)
+	}
+
+	return res, err
+}
+
+func (e *errorCachingHandler) loadErrorFromCache(ctx context.Context, key, hashedKey string) *apierror.APIError {
+	res := e.cache.GetMulti(ctx, []string{hashedKey})
+	if cached, ok := res[hashedKey]; ok {
+		var cachedError CachedError
+		if err := proto.Unmarshal(cached, &cachedError); err != nil {
+			level.Warn(e.logger).Log("msg", "unable to unmarshall cached error", "err", err)
+			return nil
+		}
+
+		if cachedError.GetKey() != key {
+			level.Debug(e.logger).Log(
+				"msg", "cached error key does not match",
+				"expected_key", key,
+				"actual_key", cachedError.GetKey(),
+				"hashed_key", hashedKey,
+			)
+			return nil
+		}
+
+		return apierror.New(apierror.Type(cachedError.Type), cachedError.Message)
+	}
+
+	return nil
+}
+
+func (e *errorCachingHandler) storeErrorToCache(key, hashedKey string, apiErr *apierror.APIError) {
+	bytes, err := proto.Marshal(&CachedError{
+		Key:     key,
+		Type:    string(apiErr.Type),
+		Message: apiErr.Message,
+	})
+
+	if err != nil {
+		level.Warn(e.logger).Log("msg", "unable to marshal cached error", "err", err)
+		return
+	}
+
+	e.cache.SetAsync(hashedKey, bytes, e.ttl)
+}
+
+func (e *errorCachingHandler) cacheKey(tenantID string, r MetricsQueryRequest) string {
+	return fmt.Sprintf("EC:%s:%s:%d:%d:%d", tenantID, r.GetQuery(), r.GetStart(), r.GetEnd(), r.GetStep())
```

Comment:
the design doc ([internal link](https://docs.google.com/document/d/1SO74280ZziBrzfzca-ZUhzkh93bI-ihEmg01TsNwIhA/edit?tab=t.0#heading=h.5sybau7waq2q)) talked about grouping the queries into time buckets. Are you thinking of doing something similar too?

Reply:
Using a bucket is less precise than start/end/step. I'm not sure if that's an appropriate thing to do since it could make a difference with respect to hitting limits.

Reply:
exact matching looked more intuitive to me too. I think I prefer it, just wanted to check with you